tools/setup-envtest/main.go

```diff
@@ -0,0 +1,273 @@
+	// flag globally.
+	envUseEnv = "ENVTEST_USE_ENV"
+)
+
+var (
+	force         = flag.Bool("force", false, "force re-downloading dependencies, even if they're already present and correct")
+	installedOnly = flag.BoolP("installed-only", "i", os.Getenv(envNoDownload) != "", "only look at installed versions -- do not query the remote API server, and error out if it would be necessary to")
+	verify        = flag.Bool("verify", true, "verify dependencies while downloading")
+	useEnv        = flag.Bool("use-env", os.Getenv(envUseEnv) != "", "whether to return the value of KUBEBUILDER_ASSETS if it's already set")
+
+	targetOS   = flag.String("os", runtime.GOOS, "os to download for (e.g. linux, darwin, for listing operations, use '*' to list all platforms)")
+	targetArch = flag.String("arch", runtime.GOARCH, "architecture to download for (e.g. amd64, for listing operations, use '*' to list all platforms)")
+
+	// printFormat is the flag value for -p, --print
+	printFormat = envp.PrintOverview
+	// zapLvl is the flag value for logging verbosity
+	zapLvl = zap.WarnLevel
+
+	binDir       = flag.String("bin-dir", "", "directory to store binary assets (default: $OS_SPECIFIC_DATA_DIR/envtest-binaries)")
+	remoteBucket = flag.String("remote-bucket", "kubebuilder-tools", "remote GCS bucket to download from")
+	remoteServer = flag.String("remote-server", "storage.googleapis.com", "remote server to query from.  You can override this if you want to run an internal storage server instead, or for testing.")
+)
+
+// TODO(directxman12): handle interrupts?
+
+// setupLogging configures a Zap logger
+func setupLogging() logr.Logger {
+	logCfg := zap.NewDevelopmentConfig()
+	logCfg.Level = zap.NewAtomicLevelAt(zapLvl)
+	zapLog, err := logCfg.Build()
+	if err != nil {
+		envp.ExitCause(1, err, "who logs the logger errors?")
+	}
+	return zapr.NewLogger(zapLog)
+}
+
+// setupEnv initializes the environment from flags.
+func setupEnv(globalLog logr.Logger, version string) *envp.Env {
+	log := globalLog.WithName("setup")
+	if *binDir == "" {
+		dataDir, err := store.DefaultStoreDir()
+		if err != nil {
+			envp.ExitCause(1, err, "unable to deterimine default binaries directory (use --bin-dir to manually override)")
+		}
+
+		*binDir = dataDir
+	}
+	log.V(1).Info("using binaries directory", "dir", *binDir)
+
+	env := &envp.Env{
+		Log:    globalLog,
+		Client: &remote.Client{
+			Log:    globalLog.WithName("storage-client"),
+			Bucket: *remoteBucket,
+			Server: *remoteServer,
+		},
+		VerifySum:     *verify,
+		ForceDownload: *force,
+		NoDownload: *installedOnly,
+		Platform: versions.PlatformItem{
+			Platform: versions.Platform{
+				OS:   *targetOS,
+				Arch: *targetArch,
+			},
+		},
+		FS: afero.Afero{Fs: afero.NewOsFs()},
+		Store: store.NewAt(*binDir),
+		Out: os.Stdout,
+	}
+
+	switch version {
+	case "", "latest":
+		env.Version = versions.LatestVersion
+	case "latest-on-disk":
+		env.Version = versions.AnyVersion
+		env.NoDownload = true
```

Comment:
Will this combination of settings match the first version found on disk, or the latest version on disk?

Reply:
latest -- we always sort by latest so that we give you the latest on disk & get deterministic behavior.

Reply:
added a comment to this effect